src/abigen/generator.ts

```diff
@@ -0,0 +1,85 @@
+import { exec, execSync } from "child_process"
+import { Artifact, Artifacts, HardhatRuntimeEnvironment } from "hardhat/types"
+import _ from "lodash"
+import { basename, resolve } from "path"
+import { existsSync } from "fs"
+import { mkdir, stat, readdir, rm, writeFile } from "fs/promises"
+
+export class Generator {
+    private abigenPath: string
+    private artifacts: Artifacts
+    private deployable: boolean
+    private dirExists: boolean
+    private outDir: string
+    private pkgName: string
+
+    constructor(hre: HardhatRuntimeEnvironment) {
+        this.abigenPath = resolve(hre.config.gobind.abigenPath)
+        this.artifacts = hre.artifacts
+        this.deployable = hre.config.gobind.deployable
+        this.outDir = resolve(hre.config.gobind.outDir)
+        this.dirExists = existsSync(this.outDir)
+        this.pkgName = _.camelCase(basename(this.outDir))
+    }
+
+    async generateAll() {
+        const names = await this.artifacts.getAllFullyQualifiedNames()
+        await this.generate(names)
+    }
+
+    async generate(artifactNames: string[]) {
+        for (const name of artifactNames) {
+            if (!this.dirExists) {
+                await mkdir(this.outDir)
+                this.dirExists = true
+            }
+
+            const artifact = await this.artifacts.readArtifact(name)
+            const abi = JSON.stringify(artifact.abi)
+            const contract = artifact.contractName
+
+            const cmdBase = `echo '${abi}' | ${this.abigenPath} --pkg ${this.pkgName} --type ${contract} --out ${this.outDir}/${contract}.go --abi -`
+            if (this.deployable) {
+                await this.generateWithBytecode(artifact, cmdBase)
+                continue
+            }
+
+            // todo: try to make exec more beautiful with args, also handle error in callback
+            exec(cmdBase, (err, _stdout, _stderr) => {
+                if (err)
+                    throw new Error(`failed to generate bindings: ${err}`)
+            })
+            // execSync(cmdBase)
+        }
+    }
+
+    async clean() {
+        if (!this.dirExists) return
```

Comment:
This is a bad idea. The user may manually delete this dir. You should check its existence right here

Reply:
It is checked in constructor, the delay between constructing and manual removing is tiny. Can we avoid additional check?

Reply:
Let's just remove the `dirExists` variable because it does not represent the actual state of the object